qiskit/transpiler/_basepasses.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""This module implements the base pass."""
+
+from abc import abstractmethod
+from collections import OrderedDict
+
+
+class MetaPass(type):
+    """
+    Enforces the creation of some fields in the pass
+    while allowing passes to override __init__
+    """
+
+    def __call__(cls, *args, **kwargs):
+        """ Called with __init__"""
+        obj = type.__call__(cls, *args, **kwargs)
+        obj._args = str(args)
+        _kwargs = OrderedDict(sorted(kwargs.items(), key=lambda t: t[0]))
+        obj._kwargs = '(' + ', '.join(["%s=%s" % (i, j) for i, j in _kwargs.items()]) + ')'
+        obj._hash = hash(obj.__repr__())
```

Comment:
This implementation of `_freeze_init_parameters` does not work with non-hashable types (and creates funny situations because `hash(True) == hash(1)`). Here is my modifications on yours:
```
    def _freeze_init_parameters(init_method, args, kwargs):
        self_guard = object()
        init_signature = signature(init_method)
        bound_signature = init_signature.bind(self_guard, *args, **kwargs)
        arguments = []
        for name,value in bound_signature.arguments.items():
            if value == self_guard:
                continue
            if isinstance(value, Hashable):
                arguments.append((name, type(value), value))
            else:
                arguments.append((name, type(value), str(value)))
        return frozenset(arguments)
```
Makes sense?